packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+        trimmedIdentifiers[i] = identifiers[i];
+      }
+      return trimmedIdentifiers;
+    } else {
+      return identifiers;
+    }
+  }
+
+  // TODO do we want to restrict permissions, or should anyone
+  // with a valid signature be able to register an attestation?
+  modifier isValidUser(address issuer, address account) {
+    require(
+      msg.sender == account ||
+        msg.sender == issuer ||
+        getAccounts().attestationSignerToAccount(msg.sender) == issuer,
+      "User does not have permission to perform this action"
+    );
+    require(!revokedSigners[msg.sender], "User has been revoked ");
+    _;
+  }
+
+  /**
+   * @notice Validates the given attestation and signature
+   * @param identifier Hash of the identifier to be attested
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account being mapped to the identifier
+   * @param issuedOn Time at which the issuer issued the attestation in Unix time 
+   * @param signer Address of the signer of the attestation
+   * @param v The recovery id of the incoming ECDSA signature
+   * @param r Output value r of the ECDSA signature
+   * @param s Output value s of the ECDSA signature
+   * @return Whether the signature is valid
+   * @dev Throws if signer is revoked
+   * @dev Throws if signer is not an authorized AttestationSigner of the issuer
+   */
+  function isValidAttestation(
+    bytes32 identifier,
+    address issuer,
+    address account,
+    uint256 issuedOn,
+    address signer,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public view returns (bool) {
+    require(!revokedSigners[signer], "Signer has been revoked");
+    require(
+      getAccounts().attestationSignerToAccount(signer) == issuer,
+      "Signer has not been authorized as an AttestationSigner by the issuer"
+    );
+    bytes32 structHash = keccak256(
+      abi.encode(EIP712_VALIDATE_ATTESTATION_TYPEHASH, identifier, issuer, account, issuedOn)
+    );
+    address guessedSigner = Signatures.getSignerOfTypedDataHash(
+      eip712DomainSeparator,
+      structHash,
+      v,
+      r,
+      s
+    );
+    return guessedSigner == signer;
+  }
+
+  /**
+   * @notice Registers an attestation with a valid signature
+   * @param identifier Hash of the identifier to be attested
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account being mapped to the identifier
+   * @param issuedOn Time at which the issuer issued the attestation in Unix time 
+   * @param signer Address of the signer of the attestation
+   * @param v The recovery id of the incoming ECDSA signature
+   * @param r Output value r of the ECDSA signature
+   * @param s Output value s of the ECDSA signature
+   * @dev Throws if sender is not the issuer, account, or an authorized AttestationSigner
+   * @dev Throws if an attestation with the same (identifier, issuer, account) already exists
+   */
+  function registerAttestation(
```

Comment:
We should also support registering attestations without an EIP-712 signature when the call comes from the issuer address (i.e. when `msg.sender == issuer`). This is valuable because it allows smart contracts to be issuers and use their own rules to define when an attestation is valid. We could also support EIP-1271 smart contract "signatures", however I would not jump on that as quickly. https://eips.ethereum.org/EIPS/eip-1271

Reply:
Ooh that's interesting! Created [a ticket](https://app.zenhub.com/workspaces/identity-backlog-board-60e61301c3dc55000e540b61/issues/celo-org/celo-monorepo/9592) for this cc @isabellewei

Reply:
oooo that is interesting :)